File: backend/content/faq_search.py
"""FAQ relevance scoring for the public API (and tests).

Clients may also filter locally using the same rules: lowercase, trim,
AND across tokens, rank question > keywords > category > answer.
"""

import re

TOKEN_RE = re.compile(r"\s+")


def tokenize_query(query):
    return [part for part in TOKEN_RE.split(str(query or "").lower().strip()) if part]


def _haystack(entry):
    keywords = entry.get("keywords") or []
    if isinstance(keywords, str):
        keywords = [part.strip() for part in keywords.split(",") if part.strip()]
    return {
        "question": str(entry.get("question") or "").lower(),
        "keywords": " ".join(str(item).lower() for item in keywords),
        "answer": str(entry.get("answer_markdown") or entry.get("answer") or "").lower(),
        "category": str(
            entry.get("category_label") or entry.get("category") or ""
        ).lower(),
    }


def score_entry(entry, tokens):
    if not tokens:
        return 0
    fields = _haystack(entry)
    blob = (
        fields["question"],
        fields["keywords"],
        fields["answer"],
        fields["category"],
    )
    for token in tokens:
        if not any(token in field for field in blob):
            return 0
    score = 0
    phrase = " ".join(tokens)
    if len(tokens) > 1 and phrase in fields["question"]:
        score += 8
    for token in tokens:
        if token in fields["question"]:
            score += 4
        if token in fields["keywords"]:
            score += 3
        if token in fields["category"]:
            score += 2
        if token in fields["answer"]:
            score += 1
    return score


def filter_faq_entries(entries, query):
    tokens = tokenize_query(query)
    items = list(entries or [])
    if not tokens:
        return items
    scored = []
    for entry in items:
        value = score_entry(entry, tokens)
        if value > 0:
            scored.append((value, entry.get("sort_order") or 0, entry))
    scored.sort(key=lambda row: (-row[0], row[1], str(row[2].get("question") or "")))
    return [row[2] for row in scored]

File: docs/static/faq-search.js
export function tokenizeQuery(query) {
  return String(query || "")
    .toLowerCase()
    .trim()
    .split(/\s+/)
    .filter(Boolean);
}

function haystack(entry) {
  const keywords = Array.isArray(entry?.keywords) ? entry.keywords : [];
  return {
    question: String(entry?.question || "").toLowerCase(),
    keywords: keywords.join(" ").toLowerCase(),
    answer: String(entry?.answer_markdown || "").toLowerCase(),
    category: String(entry?.category_label || entry?.category || "").toLowerCase(),
  };
}

export function scoreEntry(entry, tokens) {
  if (!tokens.length) return 0;
  const fields = haystack(entry);
  const blob = [fields.question, fields.keywords, fields.answer, fields.category];
  for (const token of tokens) {
    if (!blob.some((field) => field.includes(token))) return 0;
  }
  let score = 0;
  const phrase = tokens.join(" ");
  if (tokens.length > 1 && fields.question.includes(phrase)) score += 8;
  for (const token of tokens) {
    if (fields.question.includes(token)) score += 4;
    if (fields.keywords.includes(token)) score += 3;
    if (fields.category.includes(token)) score += 2;
    if (fields.answer.includes(token)) score += 1;
  }
  return score;
}

export function filterFaqEntries(entries, query) {
  const tokens = tokenizeQuery(query);
  const items = Array.isArray(entries) ? entries : [];
  if (!tokens.length) {
    return { mode: "grouped", items: [...items], query: "" };
  }
  const scored = items
    .map((entry) => ({
      entry,
      score: scoreEntry(entry, tokens),
    }))
    .filter((row) => row.score > 0)
    .sort((a, b) => {
      if (b.score !== a.score) return b.score - a.score;
      return (a.entry.sort_order || 0) - (b.entry.sort_order || 0);
    });
  return {
    mode: "search",
    items: scored.map((row) => row.entry),
    query: tokens.join(" "),
  };
}

export function searchQueryFromSearch(search) {
  const params = new URLSearchParams(String(search || "").replace(/^\?/, ""));
  return String(params.get("q") || "");
}

export function faqPathForQuery(query) {
  const trimmed = String(query || "").trim();
  if (!trimmed) return "/faq";
  return `/faq?q=${encodeURIComponent(trimmed)}`;
}

export function groupFaqByCategory(entries, categories) {
  const order = Array.isArray(categories) ? categories : [];
  const groups = new Map();
  for (const category of order) {
    groups.set(category.id, { ...category, items: [] });
  }
  for (const entry of entries || []) {
    const key = entry.category;
    if (!groups.has(key)) {
      groups.set(key, {
        id: key,
        label: entry.category_label || key,
        items: [],
      });
    }
    groups.get(key).items.push(entry);
  }
  for (const group of groups.values()) {
    group.items.sort((a, b) => (a.sort_order || 0) - (b.sort_order || 0));
  }
  return [...groups.values()].filter((group) => group.items.length);
}
